### PR TITLE
net: lib: coap_client: Improve cancel function

### PR DESCRIPTION
**Pull request by plskeggs**
Improve coap_client_cancel_requests(). Ensure it can be called from a callback. Report error to waiting callbacks. Clear active flag.

This is useful when the network becomes unavailable or prior to disconnecting in order to save power.

### CODE REVIEW

**Comment by plskeggs** [comment]
@tautologyclub I am unable to add you as a reviewer. I am modifying your cancel function so would like your input.

File: subsys/net/lib/coap/coap_client.c
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:836]: Not sure I understand this part, what if the mutex was locked already, why are we just ignoring this fact?

Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c]: Removed. I think locking isn't necessary here anyway.

**Comment by tautologyclub** [comment]
I def have opinions here, but it'll have to wait until I'm not mega sick anymore, ie a day or two

File: subsys/net/lib/coap/coap_client.c
Comment (tautologyclub) [inline subsys/net/lib/coap/coap_client.c:839]: Request force-cancelled seems different to me than a "regular" coap timeout. Isn't `-errno` codes already used for stuff like this? How about simply `-ECANCELED`?

File: subsys/net/lib/coap/coap_client.c
Comment (tautologyclub) [reply subsys/net/lib/coap/coap_client.c:779]: maybe it's just your comment being slightly misleading because I suspect you mean "request cancelled asynchronously" or whatever, but I'm not sure

**Comment by tautologyclub** [comment]
In general it seems like a strict improvement so LGTM if you've tested it thoroughly

Though I must once again note that my PR should not have been merged. I planned to make a new PR, but I'll hold off until this one is sorted...

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c:779]: I meant that we allow the user to cancel it asynchronously, including from their callback, by calling `coap_client_cancel_requests()`.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c]: I also removed the code that touches the pending list. It is enough to just mark the request as not ongoing.

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c]: Done

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c]: Rewrote the comment.

File: subsys/net/lib/coap/coap_client.c
Comment (glarsennordic) [inline subsys/net/lib/coap/coap_client.c:27]: While we are in the area, can we please add some comments to `send_request`, `send_ack` and `handle_request` indicating the intended expected return values?

For `send_request` and `send_ack` it is pretty simple, they return a negative value on error, and a positive value indicating total number of bytes sent on success.


But `handle_request` is very difficult to untangle due to its inconsistent and undocumented control flow. It actually has three different non-negative return values (0, 1, and under certain conditions, the exact value that `send_ack` returned, verbatim), which are inconsistent enough as to not even be useful to a potential caller. And the `fail` block is, IMO, misleadingly named.

Therefore, could we additionally please:
1. Make those return values consistent
2. Add a comment to the `send_ack` control flow indicating that the passthrough into the `fail` block is ***intentional***, and is (I presume) done because sending an ack indicates that the request has completed (if it is non-OBSERVE)
3. Add a comment to the `send_request` control flow indicating that we return early with `1` because (I presume) the overall request is still ongoing due to the followup request?
4. Possibly rename the `fail` block? -- That is, if you can think of a better name. Maybe just `exit` or `cleanup` or `conclude` or `finish`? IDK... Since this block now also handles OBSERVE requests, its kinda difficult to name properly. Maybe this indicates the function should be broken down differently, but I won't ask you to totally refactor it 😅

File: subsys/net/lib/coap/coap_client.c
Comment (glarsennordic) [inline subsys/net/lib/coap/coap_client.c:331]: Perhaps `send_mutex` instead of just `mutex`. It might make finding the source of this message a little easier

File: subsys/net/lib/coap/coap_client.c
Comment (glarsennordic) [inline subsys/net/lib/coap/coap_client.c:396]: Would an else block printing a debug message stating there was an error during handling of callback make sense? I worry a little about potentially causing dropped error callbacks. But I don't understand this client well enough to know if that concern is reasonable

**Review by glarsennordic — COMMENTED** [review]
I noticed a potential bug in `handle_request` which was introduced when https://github.com/zephyrproject-rtos/zephyr/pull/72159 was unintentionally merged.

Here is my reasoning, please correct me if I got any of this wrong:

First, check out this diff from that PR:

```diff
-internal_req->request_ongoing = false;
+if (ret != 0 || !coap_request_is_observe(&internal_req->request)) {
+	internal_req->request_ongoing = false;
+}
```

What is the intended purpose for checking `ret != 0`? Negative `ret` indicates error. What does positive `ret` indicate?

The only two functions that `handle_response` calls that can return positive return values are `send_request` and `send_ack`, both of which use positive values to indicate the number of bytes sent.

But for `send_request`, this gets converted to an early `return 1`, whereas for `send_ack`, this value ***might*** (under the right conditions) be returned directly via the `fail` block?

Presumably, the difference here is that `send_request` results in an as-yet ongoing request, whereas `send_ack` completes the request?

So, originally, the `fail` block was essentially handling all scenarios where the request finishes -- either due to error, or simply having completed successfully. 

*This is where we get to the bug*:

What if:
1. `send_ack` succeeds
2. `block_option` happens to be non-positive
3. `blockwise_transfer` happens to be false, or `last_block` happens to be true
4. The request is of OBSERVE type

If those four conditions are met, then the request will be incorrectly marked as no-longer ongoing, won't it? Or is this just not possible? Or is it intended, for some reason I don't understand?

File: subsys/net/lib/coap/coap_client.c
Comment (glarsennordic) [reply subsys/net/lib/coap/coap_client.c:27]: Discussed offline, these readability tweaks are out of scope for what we are working on. But @tautologyclub you may consider them my recommendations to you for when you do that followup PR you mentioned

**Comment by glarsennordic** [comment]
My comment about the possible bug introduced by https://github.com/zephyrproject-rtos/zephyr/pull/72159 is out of scope for this PR, my bad!

I'll instead copy the comment to that PR and discuss with @tautologyclub about whether it is genuinely a problem, and what to do about it

File: subsys/net/lib/coap/coap_client.c
Comment (glarsennordic) [reply subsys/net/lib/coap/coap_client.c:27]: That being said, @juhaylinen What do you think of these suggestions? Shall I file a github issue about them?

File: subsys/net/lib/coap/coap_client.c
Comment (plskeggs) [reply subsys/net/lib/coap/coap_client.c]: It will only skip reporting when the report is due to a call to the cancel function from the same callback that would receive the error notification. But I went ahead and made that explicit.

**Review by rlubos — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Pull request merged**
